app/controllers/shipping_api_controller.rb

```diff
@@ -0,0 +1,64 @@
+class ShippingApiController < ApplicationController
+  skip_before_filter :verify_authenticity_token, only: [:rates, :create_log]
+  before_action :set_shipping_rate, only: :rates
+
+  def rates
+    if @shipping_rate.valid?
+      estimates = @shipping_rate.ups_rates + @shipping_rate.usps_rates
+      render json: estimates, status: :ok # TODO: add status code tests
+    else
+      render json: @shipping_rate.errors, status: :bad_request # TODO: add status code tests
+
+    end
+  end
+
+  def create_log
+    log = Log.new(log_params)
+
+    if log.save
+      render json: log, status: :ok # 200
+    else
+      render json: log.errors, status: :bad_request # 400
+    end
+  end
+
+  private
+
+  def origin
+    ShippingLocation.new(origin_params)
+  end
+
+  def origin_params
+    params.require(:origin).permit(:city, :state, :zip, :country)
+  end
+
+  def destination
+    ShippingLocation.new(destination_params)
+  end
+
+  def destination_params
+    params.require(:destination).permit(:city, :state, :zip, :country)
+  end
+
+  def package
+    unless package_params[:weight].nil? || package_params[:dimensions].nil? || package_params[:units].nil?
```

Comment:
Generally I try to only use `unless` when there's a single test in the statement. Otherwise it gets kind of hard to figure out. If you want to avoid a double negative construct like `!package_params[:weight].nil?` you can use `package_params[:weight].present?`.